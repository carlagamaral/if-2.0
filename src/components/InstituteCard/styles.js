import styled from 'styled-components';

export const WrapperInstituteCard = styled.div`
	display: flex;
	justify-content: space-between;
	align-items: center;
	height: '${(props) => props.cardHeight}px';
	width: ${({ fromMap }) => (fromMap === true ? '130%' : '65%')};
	border-radius: ${({ fromMap }) => (fromMap === true ? '15px' : '')};
	background-color: ${({ fromMap }) => fromMap === true && 'white'};
	line-height: 25px;

	${({ fromMap }) =>
		fromMap === true &&
		`
    box-shadow: 0 0  0 2px rgb(255,255,255),
                0.3em 0.3em 1em rgba(0,0,0,0.3);
`}
`;

export const FullCard = styled.div`
	display: flex;
	justify-content: center;
	width: 100%;
	align-items: center;
	height: ${({ fromMap }) => (fromMap === true ? '200px' : '')};
	padding-bottom: ${({ fromMap }) => (fromMap === true ? '20px' : '')};
`;

export const Infos = styled.div`
	width: 80%;
`;

export const Title = styled.span`
	padding-top: 20px;
	font-size: 20px;
	display: flex;
	align-items: flex-start;
`;
