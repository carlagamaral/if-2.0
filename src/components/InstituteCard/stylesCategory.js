import styled from 'styled-components';

export const CategoryWrapper = styled.div`
	display: flex;
	align-items: center;
	font-size: 14px;

	.letter {
		border-radius: 15px;
		font-size: ${(props) => props.letterFontSize}px;
		background-color: ${(props) => props.letterColor};
		color: transparent;
		width: 40px;
		height: 40px;
		margin-right: 10px;
	}
`;

export const Bottom = styled.div`
	font-size: 16px;
	font-weight: 300;
	line-height: 20px;

	span {
		display: ${({ fromMap }) => (fromMap ? '' : 'flex')};
		align-items: ${({ fromMap }) => (fromMap ? '' : 'flex-start')};
	}
`;

export const NumberAndPercentage = styled.div`
	display: flex;
	justify-content: space-around;

	span + span {
		font-weight: bold;
	}
`;
