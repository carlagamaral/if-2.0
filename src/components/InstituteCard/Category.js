import React from 'react';
import PropTypes from 'prop-types';
import { CategoryWrapper, Bottom, NumberAndPercentage } from './stylesCategory';

const Category = ({
	counter,
	description,
	letterColor,
	letterFontSize,
	percentagePlaces,
	testId,
	total,
	fromMap,
}) => {
	// We have this weird (total || 1) to account for the total being 0, so we just divide 0 by 1,
	// which will have a percentage that makes sense (0%)
	const percentage = ((counter / (total || 1)) * 100).toFixed(percentagePlaces);

	return (
		<CategoryWrapper
			data-testid={testId}
			letterColor={letterColor}
			letterFontSize={letterFontSize}
		>
			<span className="letter" data-testid={`${testId}-letter`}>
				{description === 'Artes Visuais' ? 'AV' : description.toUpperCase()[0]}
			</span>
			<Bottom>
				<span data-testid={`${testId}-description`} fromMap={fromMap}>
					{description}
				</span>
				<NumberAndPercentage>
					<span data-testid={`${testId}-counter`}>{counter} -&nbsp;</span>
					<span data-testid={`${testId}-counter-percentage`}>
						{percentage}%
					</span>
				</NumberAndPercentage>
			</Bottom>
		</CategoryWrapper>
	);
};

Category.propTypes = {
	counter: PropTypes.number.isRequired,
	description: PropTypes.string.isRequired,
	fromMap: PropTypes.bool,
	letterColor: PropTypes.string,
	letterFontSize: PropTypes.number,
	percentagePlaces: PropTypes.number,
	testId: PropTypes.string.isRequired,
	total: PropTypes.number.isRequired,
};

Category.defaultProps = {
	fromMap: false,
	letterColor: 'black',
	letterFontSize: 40,
	percentagePlaces: 2,
};

export default Category;
