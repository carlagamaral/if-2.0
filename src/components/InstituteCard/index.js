import React from 'react';
import PropTypes from 'prop-types';
import Category from './Category';
import { WrapperInstituteCard, FullCard, Title, Infos } from './styles';

const InstituteCard = ({
	height,
	width,
	name,
	imagesHeight,
	imagesWidth,
	categories,
	fromMap,
}) => {
	const total = categories.reduce((acc, { counter }) => acc + counter, 0);

	return (
		<WrapperInstituteCard
			cardHeight={height}
			cardWidth={width}
			data-testid="wrapper-institute-card"
			fromMap={fromMap}
		>
			<FullCard fromMap={fromMap}>
				<Infos>
					<Title data-testid="header-name">
						{name === 'Total' ? `${total} docentes` : `${name}`}
					</Title>
					<div className="images" data-testid="categories">
						{categories.map((category) => (
							<Category
								counter={category.counter}
								description={category.description}
								imagesHeight={imagesHeight}
								imagesWidth={imagesWidth}
								key={category.description}
								letterColor={category.color}
								testId={`category-${category.description}`}
								total={total}
							/>
						))}
					</div>
				</Infos>
			</FullCard>
		</WrapperInstituteCard>
	);
};

InstituteCard.propTypes = {
	categories: PropTypes.arrayOf(
		PropTypes.shape({
			counter: PropTypes.number,
			description: PropTypes.string,
			name: PropTypes.string,
		})
	),
	fromMap: PropTypes.bool,
	height: PropTypes.number,
	imagesHeight: PropTypes.number,
	imagesWidth: PropTypes.number,
	name: PropTypes.string,
	width: PropTypes.string,
};

InstituteCard.defaultProps = {
	categories: [
		{
			counter: 0,
			description: '',
			name: '',
		},
	],
	fromMap: false,
	height: 220,
	imagesHeight: 50,
	imagesWidth: 50,
	name: 'Nome',
	width: '350px',
};

export default InstituteCard;
