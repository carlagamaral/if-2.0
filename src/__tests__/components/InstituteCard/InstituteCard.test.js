/* eslint-disable function-paren-newline */
import 'jest-dom/extend-expect'
import { cleanup, waitForElement, render } from 'react-testing-library'
import React from 'react'
import InstituteCard from '../../../components/InstituteCard'

describe( 'InstituteCard tests ', () => {
  const name = 'Naruto'
  const categories = [
    { 'background-color': '#123456', 'counter': 2, 'description': 'desc1', 'name': 'desc1' },
    { 'background-color': '#123455', 'counter': 4, 'description': 'desc2', 'name': 'desc2' },
    { 'background-color': '#123433', 'counter': 5, 'description': 'desc3', 'name': 'desc3' },
  ]


  afterEach( () => {
    cleanup()
    jest.clearAllMocks()
  } )

  test( 'Name is rendered on InstituteCard', async() => {
    const { container, getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )
    const header = await waitForElement( () => getByTestId( 'header-name' ) )

    expect( header.innerHTML ).toEqual( name )
    expect( container ).toMatchSnapshot()
  } )

  test( 'Every category is rendered', async() => {
    const { container, getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )
    const categoriesContainer = await waitForElement( () => getByTestId( 'categories' ) )

    expect( categoriesContainer.childElementCount ).toEqual( categories.length )
    expect( container ).toMatchSnapshot()
  } )

  test( 'Every category is rendered with the right description letter', async() => {
    const { getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )

    for ( const category of categories ) {
      const categoryContainer = await waitForElement( () => getByTestId( `category-${category.description}-letter` ) ) // eslint-disable-line no-await-in-loop

      expect( categoryContainer.innerHTML ).toEqual( category.description.toUpperCase()[ 0 ] )
      expect( categoryContainer ).toHaveStyle( `background-color: ${category.color}` )
    }
  } )

  test( 'Every category is rendered with the right description', async() => {
    const { getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )

    for ( const category of categories ) {
      const categoryContainer = await waitForElement( () => getByTestId( `category-${category.description}-description` ) ) // eslint-disable-line no-await-in-loop

      expect( categoryContainer.innerHTML ).toEqual( category.description )
    }
  } )

  test( 'Every category is rendered with the right counter', async() => {
    const { getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )

    for ( const category of categories ) {
      const categoryContainer = await waitForElement( () => getByTestId( `category-${category.description}-counter` ) ) // eslint-disable-line no-await-in-loop

      expect( categoryContainer.innerHTML ).toEqual( category.counter.toString() )
    }
  } )

  test( 'Every category is rendered with the right percentage', async() => {
    const { getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )

    const total = categories.reduce( ( acc, category ) => acc + category.counter, 0 )
    for ( const category of categories ) {
      const categoryContainer = await waitForElement( () => getByTestId( `category-${category.description}-counter-percentage` ) ) // eslint-disable-line no-await-in-loop

      expect( categoryContainer.innerHTML ).toEqual( `${( ( category.counter / total ) * 100 ).toFixed( 2 )}%` )
    }
  } )

  test( 'The categories counter total is correctly displayed', async() => {
    const { getByTestId } = render(
      <InstituteCard
        categories={ categories }
        name={ name }
      />
    )
    const totalSpan = await waitForElement( () => getByTestId( 'total' ) )

    const sumCounters = categories.reduce( ( acc, { counter } ) => acc + counter, 0 )
    expect( totalSpan.innerHTML ).toEqual( `Total: ${sumCounters}` )
  } )
} )
