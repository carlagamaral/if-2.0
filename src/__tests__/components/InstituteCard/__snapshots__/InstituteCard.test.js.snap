// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstituteCard tests  Every category is rendered 1`] = `
<div>
  <div
    class="sc-AxirZ gsECqv"
    data-testid="wrapper-institute-card"
  >
    <div
      class="header"
    >
      <h2
        data-testid="header-name"
      >
        Naruto
      </h2>
    </div>
    <div
      class="images"
      data-testid="categories"
    >
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc1"
      >
        <span
          class="letter"
          data-testid="category-desc1-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc1-description"
        >
          desc1
        </span>
        <span
          class="counter"
          data-testid="category-desc1-counter"
        >
          2
        </span>
        <span
          class="counter"
          data-testid="category-desc1-counter-percentage"
        >
          18.18
          %
        </span>
      </div>
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc2"
      >
        <span
          class="letter"
          data-testid="category-desc2-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc2-description"
        >
          desc2
        </span>
        <span
          class="counter"
          data-testid="category-desc2-counter"
        >
          4
        </span>
        <span
          class="counter"
          data-testid="category-desc2-counter-percentage"
        >
          36.36
          %
        </span>
      </div>
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc3"
      >
        <span
          class="letter"
          data-testid="category-desc3-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc3-description"
        >
          desc3
        </span>
        <span
          class="counter"
          data-testid="category-desc3-counter"
        >
          5
        </span>
        <span
          class="counter"
          data-testid="category-desc3-counter-percentage"
        >
          45.45
          %
        </span>
      </div>
    </div>
    <div
      class="total"
    >
      <span
        data-testid="total"
      >
        Total: 
        11
      </span>
    </div>
  </div>
</div>
`;

exports[`InstituteCard tests  Name is rendered on InstituteCard 1`] = `
<div>
  <div
    class="sc-AxirZ gsECqv"
    data-testid="wrapper-institute-card"
  >
    <div
      class="header"
    >
      <h2
        data-testid="header-name"
      >
        Naruto
      </h2>
    </div>
    <div
      class="images"
      data-testid="categories"
    >
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc1"
      >
        <span
          class="letter"
          data-testid="category-desc1-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc1-description"
        >
          desc1
        </span>
        <span
          class="counter"
          data-testid="category-desc1-counter"
        >
          2
        </span>
        <span
          class="counter"
          data-testid="category-desc1-counter-percentage"
        >
          18.18
          %
        </span>
      </div>
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc2"
      >
        <span
          class="letter"
          data-testid="category-desc2-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc2-description"
        >
          desc2
        </span>
        <span
          class="counter"
          data-testid="category-desc2-counter"
        >
          4
        </span>
        <span
          class="counter"
          data-testid="category-desc2-counter-percentage"
        >
          36.36
          %
        </span>
      </div>
      <div
        class="sc-AxjAm gUBFqK"
        data-testid="category-desc3"
      >
        <span
          class="letter"
          data-testid="category-desc3-letter"
        >
          D
        </span>
        <span
          class="description"
          data-testid="category-desc3-description"
        >
          desc3
        </span>
        <span
          class="counter"
          data-testid="category-desc3-counter"
        >
          5
        </span>
        <span
          class="counter"
          data-testid="category-desc3-counter-percentage"
        >
          45.45
          %
        </span>
      </div>
    </div>
    <div
      class="total"
    >
      <span
        data-testid="total"
      >
        Total: 
        11
      </span>
    </div>
  </div>
</div>
`;
